gc/base/SparseAddressOrderedFixedSizeDataPool.cpp

```diff
@@ -0,0 +1,412 @@
+ * the terms of the Eclipse Public License 2.0 which accompanies this
+ * distribution and is available at https://www.eclipse.org/legal/epl-2.0/
+ * or the Apache License, Version 2.0 which accompanies this distribution and
+ * is available at https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * This Source Code may also be made available under the following
+ * Secondary Licenses when the conditions for such availability set
+ * forth in the Eclipse Public License, v. 2.0 are satisfied: GNU
+ * General Public License, version 2 with the GNU Classpath
+ * Exception [1] and GNU General Public License, version 2 with the
+ * OpenJDK Assembly Exception [2].
+ *
+ * [1] https://www.gnu.org/software/classpath/license.html
+ * [2] http://openjdk.java.net/legal/assembly-exception.html
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+ *******************************************************************************/
+
+#include "omrcomp.h"
+#include "modronbase.h"
+#include "ModronAssertions.h"
+#include "SparseVirtualMemory.hpp"
+#include "SparseAddressOrderedFixedSizeDataPool.hpp"
+#include "Heap.hpp"
+
+MM_SparseAddressOrderedFixedSizeDataPool *
+MM_SparseAddressOrderedFixedSizeDataPool::newInstance(MM_EnvironmentBase *env, void *sparseHeapBase, uintptr_t sparseDataPoolSize)
+{
+	MM_SparseAddressOrderedFixedSizeDataPool *sparseDataPool;
+
+	sparseDataPool = (MM_SparseAddressOrderedFixedSizeDataPool *)env->getForge()->allocate(sizeof(MM_SparseAddressOrderedFixedSizeDataPool), OMR::GC::AllocationCategory::FIXED, OMR_GET_CALLSITE());
+	if (sparseDataPool) {
+		Trc_MM_SparseAddressOrderedFixedSizeDataPool_allocation_success(sparseHeapBase, (void *)sparseDataPoolSize);
+		sparseDataPool = new(sparseDataPool) MM_SparseAddressOrderedFixedSizeDataPool(env, sparseDataPoolSize);
+		if (!sparseDataPool->initialize(env, sparseHeapBase)) {
+			sparseDataPool->kill(env);
+			sparseDataPool = NULL;
+		}
+	} else {
+		Trc_MM_SparseAddressOrderedFixedSizeDataPool_allocation_failure(sparseHeapBase, sparseDataPoolSize);
+	}
+
+	return sparseDataPool;
+}
+
+uintptr_t
+MM_SparseAddressOrderedFixedSizeDataPool::entryHash(void *entry, void *userData)
+{
+	return (uintptr_t)((SparseDataTableEntry *)entry)->dataPtr;
+}
+
+uintptr_t
+MM_SparseAddressOrderedFixedSizeDataPool::entryEquals(void *leftEntry, void *rightEntry, void *userData)
+{
+	uintptr_t lhs = (uintptr_t)((SparseDataTableEntry *)leftEntry)->dataPtr;
+	uintptr_t rhs = (uintptr_t)((SparseDataTableEntry *)rightEntry)->dataPtr;
+	return (lhs == rhs);
+}
+
+bool
+MM_SparseAddressOrderedFixedSizeDataPool::initialize(MM_EnvironmentBase *env, void *sparseHeapBase)
+{
+	OMRPORT_ACCESS_FROM_OMRPORT(env->getPortLibrary());
+
+	/* Initialize _freeListPool J9Pool */
+	_freeListPool = pool_new(sizeof(MM_SparseHeapLinkedFreeHeader), 6, 0, 0, OMR_GET_CALLSITE(), OMRMEM_CATEGORY_MM, POOL_FOR_PORT(OMRPORTLIB));
```

Comment:
Would you please define preprocessor constant for number of pool elements and pass it as parameter (instead of `6`)?